src/components/file/media/media-detector.service.ts

```diff
@@ -0,0 +1,83 @@
+import ffprobeBinary from '@ffprobe-installer/ffprobe';
+import { Injectable } from '@nestjs/common';
+import execa from 'execa';
+import { FFProbeResult } from 'ffprobe';
+import { imageSize } from 'image-size';
+import { ISize as ImageSize } from 'image-size/dist/types/interface';
+import { Readable } from 'stream';
+import { Except } from 'type-fest';
+import { Downloadable } from '../dto';
+import { AnyMedia, Media } from './media.dto';
+
+@Injectable()
+export class MediaDetector {
+  async detect(
+    file: Downloadable<{ mimeType: string }>,
+  ): Promise<Except<AnyMedia, Exclude<keyof Media, '__typename'>> | null> {
+    if (file.mimeType.startsWith('image/')) {
+      const buffer = await file.download();
+
+      let size: ImageSize = { width: 0, height: 0 };
+      try {
+        size = imageSize(buffer);
+      } catch (e) {
+        // ignore
+      }
+
+      return {
+        __typename: 'Image',
+        dimensions: {
+          width: size.width ?? 0,
+          height: size.height ?? 0,
+        },
+      };
+    }
+
+    const isAudio = file.mimeType.startsWith('audio/');
```

Comment:
Will this sort of audio, video, image strategy be repeated anywhere else? IF so it might be worth assigning 'audio/', 'video/' etc to an enum rather than hardcoding them

Reply:
I doubt it. This is the only spot we have logic based on mime types. But definitely something to keep in mind for future.
Depending on the future need, the media type could be check instead, which is a string literal union.
```ts
media.__typename === 'Audio'
```